File: frontend/src/types/actions.ts
import type { NodeData, EdgeData } from './graph';

export type AddNodeAction = {
  op: 'add_node';
  id: string;
  type: string;
  position: { x: number; y: number };
  data: NodeData;
};

export type RemoveNodeAction = {
  op: 'remove_node';
  id: string;
};

export type UpdateNodeAction = {
  op: 'update_node';
  id: string;
  data: Partial<NodeData>;
};

export type MoveNodeAction = {
  op: 'move_node';
  id: string;
  position: { x: number; y: number };
};

export type AddEdgeAction = {
  op: 'add_edge';
  id: string;
  source: string;
  target: string;
  data?: EdgeData;
};

export type RemoveEdgeAction = {
  op: 'remove_edge';
  id: string;
};

export type UpdateEdgeAction = {
  op: 'update_edge';
  id: string;
  data: Partial<EdgeData>;
};

export type GraphAction =
  | AddNodeAction
  | RemoveNodeAction
  | UpdateNodeAction
  | MoveNodeAction
  | AddEdgeAction
  | RemoveEdgeAction
  | UpdateEdgeAction;

export type AIResponse = {
  thought_process: string;
  actions: GraphAction[];
  summary: string;
};

export type ChatMessage = {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  thought_process?: string;
  timestamp: string;
};

// ── Streaming types ──────────────────────────────────────

export type StreamEvent =
  | { type: 'token'; token: string }
  | { type: 'tool_start'; name: string; input: Record<string, unknown> }
  | { type: 'tool_end'; name: string; output: Record<string, unknown> }
  | { type: 'done'; response: AIResponse }
  | { type: 'error'; message: string };

export type ToolCallState = {
  name: string;
  status: 'running' | 'done';
  input: Record<string, unknown>;
  output?: Record<string, unknown>;
};

File: backend/app/models/actions.py
from __future__ import annotations
from typing import Annotated, Literal, Optional, Union
from pydantic import BaseModel, Discriminator, Tag

from .graph import NodeData, EdgeData, Position


class AddNodeAction(BaseModel):
    op: Literal["add_node"]
    id: str
    type: str
    position: Position
    data: NodeData


class RemoveNodeAction(BaseModel):
    op: Literal["remove_node"]
    id: str


class UpdateNodeAction(BaseModel):
    op: Literal["update_node"]
    id: str
    data: dict  # Partial<NodeData>


class MoveNodeAction(BaseModel):
    op: Literal["move_node"]
    id: str
    position: Position


class AddEdgeAction(BaseModel):
    op: Literal["add_edge"]
    id: str
    source: str
    target: str
    data: Optional[EdgeData] = None


class RemoveEdgeAction(BaseModel):
    op: Literal["remove_edge"]
    id: str


class UpdateEdgeAction(BaseModel):
    op: Literal["update_edge"]
    id: str
    data: dict  # Partial<EdgeData>


def _get_discriminator_value(v: dict) -> str:
    if isinstance(v, dict):
        return v.get("op", "")
    return getattr(v, "op", "")


GraphAction = Annotated[
    Union[
        Annotated[AddNodeAction, Tag("add_node")],
        Annotated[RemoveNodeAction, Tag("remove_node")],
        Annotated[UpdateNodeAction, Tag("update_node")],
        Annotated[MoveNodeAction, Tag("move_node")],
        Annotated[AddEdgeAction, Tag("add_edge")],
        Annotated[RemoveEdgeAction, Tag("remove_edge")],
        Annotated[UpdateEdgeAction, Tag("update_edge")],
    ],
    Discriminator(_get_discriminator_value),
]


class AIResponse(BaseModel):
    thought_process: str
    actions: list[GraphAction]
    summary: str
